# ML-04 — Search Intelligence Data Contract

[![Open In Colab](https://colab.research.google.com/assets/colab-badge.svg)](https://colab.research.google.com/github/ishuum13-star/flyrank-ml-internship-v3/blob/main/work/notebooks/w03_data_contract.ipynb?flush_cache=true)

This skeleton is yours to fill. Work the sections **in order** — each one has a one-line hint. Simple words, honest numbers.

> Working with an AI assistant? Tell it to read `skills/README.md` first and load the one skill this assignment names on its card.

Cell 1 (Setup):

In [52]:
!pip install -q duckdb

import duckdb
from google.colab import userdata

HF_TOKEN = userdata.get('HF_TOKEN')

con = duckdb.connect()
con.execute(f"CREATE SECRET (TYPE huggingface, TOKEN '{HF_TOKEN}')")

BASE = "hf://datasets/FlyRank/internship-warehouse"
print("Connected. Ready to explore.")

Connected. Ready to explore.


Cell 2 (File list):

In [53]:
files = con.sql(f"SELECT * FROM glob('{BASE}/**') LIMIT 50").df()
print(files)

                                                 file
0   hf://datasets/FlyRank/internship-warehouse/.gi...
1   hf://datasets/FlyRank/internship-warehouse/REA...
2   hf://datasets/FlyRank/internship-warehouse/dim...
3   hf://datasets/FlyRank/internship-warehouse/dim...
4   hf://datasets/FlyRank/internship-warehouse/fac...
5   hf://datasets/FlyRank/internship-warehouse/fac...
6   hf://datasets/FlyRank/internship-warehouse/fac...
7   hf://datasets/FlyRank/internship-warehouse/fac...
8   hf://datasets/FlyRank/internship-warehouse/fac...
9   hf://datasets/FlyRank/internship-warehouse/fac...
10  hf://datasets/FlyRank/internship-warehouse/fac...
11  hf://datasets/FlyRank/internship-warehouse/fac...
12  hf://datasets/FlyRank/internship-warehouse/fac...
13  hf://datasets/FlyRank/internship-warehouse/fac...
14  hf://datasets/FlyRank/internship-warehouse/fac...
15  hf://datasets/FlyRank/internship-warehouse/fac...
16  hf://datasets/FlyRank/internship-warehouse/fac...
17  hf://datasets/FlyRank/in

In [54]:
for f in files["file"].tolist():
    print(f)

hf://datasets/FlyRank/internship-warehouse/.gitattributes
hf://datasets/FlyRank/internship-warehouse/README.md
hf://datasets/FlyRank/internship-warehouse/dim_clients.parquet
hf://datasets/FlyRank/internship-warehouse/dim_content.parquet
hf://datasets/FlyRank/internship-warehouse/fact_content_daily_performance/month=2025-01/data_0.parquet
hf://datasets/FlyRank/internship-warehouse/fact_content_daily_performance/month=2025-02/data_0.parquet
hf://datasets/FlyRank/internship-warehouse/fact_content_daily_performance/month=2025-03/data_0.parquet
hf://datasets/FlyRank/internship-warehouse/fact_content_daily_performance/month=2025-04/data_0.parquet
hf://datasets/FlyRank/internship-warehouse/fact_content_daily_performance/month=2025-05/data_0.parquet
hf://datasets/FlyRank/internship-warehouse/fact_content_daily_performance/month=2025-06/data_0.parquet
hf://datasets/FlyRank/internship-warehouse/fact_content_daily_performance/month=2025-07/data_0.parquet
hf://datasets/FlyRank/internship-warehouse

In [55]:
# Peek at schema + a few rows from the mid-panel month
daily = con.sql(f"SELECT * FROM '{BASE}/fact_content_daily_performance/month=2026-03/data_0.parquet' LIMIT 5").df()
print(daily.columns.tolist())
daily

['report_date', 'client_hash_id', 'content_hash_id', 'client_has_gsc', 'client_has_ga4', 'gsc_data_available', 'ga4_data_available', 'gsc_impressions', 'gsc_clicks', 'gsc_sum_position', 'gsc_avg_position', 'ga4_pageviews', 'ga4_sessions', 'ga4_users', 'ga4_engaged_sessions', 'ga4_total_engagement_sec', 'sessions_organic', 'sessions_direct', 'sessions_referral', 'sessions_social', 'sessions_paid', 'sessions_ai', 'ai_chatgpt', 'ai_perplexity', 'ai_gemini', 'ai_copilot', 'ai_claude', 'ai_meta', 'ai_other', 'scroll_events', 'month']


,report_date,client_hash_id,content_hash_id,client_has_gsc,client_has_ga4,gsc_data_available,ga4_data_available,gsc_impressions,gsc_clicks,gsc_sum_position,...,sessions_ai,ai_chatgpt,ai_perplexity,ai_gemini,ai_copilot,ai_claude,ai_meta,ai_other,scroll_events,month
0,2026-03-01,client_73cda7b4e4f265ea,content_b7e512995f79d5a6,True,False,True,<NA>,20,0,67,...,<NA>,<NA>,<NA>,<NA>,<NA>,<NA>,<NA>,<NA>,<NA>,2026-03
1,2026-03-01,client_73cda7b4e4f265ea,content_05597932fe4da067,True,False,True,<NA>,1,0,0,...,<NA>,<NA>,<NA>,<NA>,<NA>,<NA>,<NA>,<NA>,<NA>,2026-03
2,2026-03-01,client_73cda7b4e4f265ea,content_7a105f548d9c6916,True,False,True,<NA>,125,1,616,...,<NA>,<NA>,<NA>,<NA>,<NA>,<NA>,<NA>,<NA>,<NA>,2026-03
3,2026-03-01,client_73cda7b4e4f265ea,content_905aa32a0230694e,True,False,True,<NA>,7,0,28,...,<NA>,<NA>,<NA>,<NA>,<NA>,<NA>,<NA>,<NA>,<NA>,2026-03
4,2026-03-01,client_73cda7b4e4f265ea,content_a3ea9792f793ec72,True,False,True,<NA>,11,0,25,...,<NA>,<NA>,<NA>,<NA>,<NA>,<NA>,<NA>,<NA>,<NA>,2026-03


In [56]:
content = con.sql(f"SELECT * FROM '{BASE}/dim_content.parquet' LIMIT 3").df()
print("dim_content columns:", content.columns.tolist())

clients = con.sql(f"SELECT * FROM '{BASE}/dim_clients.parquet' LIMIT 3").df()
print("dim_clients columns:", clients.columns.tolist())

dim_content columns: ['client_hash_id', 'content_hash_id', 'keyword_hash_id', 'url_hash_id', 'keyword_char_count', 'keyword_token_count', 'url_char_count', 'content_created_date', 'content_updated_date', 'content_type', 'search_volume', 'competition', 'competition_level', 'cpc', 'main_intent', 'backlinks', 'category_count', 'keyword_created_date', 'provider_used', 'model_used', 'char_count', 'word_count', 'last_optimized_date', 'optimization_eligible_date', 'is_published', 'is_deleted']
dim_clients columns: ['client_hash_id', 'is_active', 'has_gsc_access', 'has_ga4_access', 'access_profile', 'client_created_date', 'client_updated_date', 'gsc_data_start', 'ga4_data_start']


## 1. Unit of analysis + time window

*One row = one what, over which dates? State it, then verify it below.*

One row = one content page (content_hash_id) belonging to one client (client_hash_id),
aggregated over report_date within a single calendar month. Time window: month=2026-03
(March 2026) — a mid-panel month, not the final month (June 2026 is reserved as the
sealed test month, per the warehouse's own _sample file). The raw fact_content_daily_performance
table is daily-grain (one row = one content page-day); I aggregate it to page-month grain
(sum impressions/clicks, average position) before building my lane's feature frame.

## 2. Fields: feature / label / context / excluded

*Sort every field you plan to touch into these four buckets. Excluded needs a why.*

FEATURE: dim_content.content_type, main_intent, word_count, char_count, search_volume,
competition, cpc, backlinks; aggregated from fact — sum(gsc_impressions), sum(gsc_clicks),
avg(gsc_avg_position) for March.

LABEL/PROXY: declining = 1 if sum(gsc_impressions) in March 2026 < sum(gsc_impressions) in
February 2026, for the same content_hash_id. A rule-derived proxy, not a directly observed
"this page failed" outcome — it depends on GSC impressions being tracked consistently both months.

CONTEXT (used for joins/filters, not fed as a feature): report_date, client_hash_id,
content_hash_id, month, content_created_date, last_optimized_date, client_has_gsc/ga4_data_available.

EXCLUDED: content_hash_id / client_hash_id / keyword_hash_id / url_hash_id — salted hash
identifiers; excluding them as features stops the model from memorizing specific pages
instead of learning a generalizable pattern. Also excluded: provider_used / model_used —
metadata about how content was produced, not something I want driving prioritization decisions
this early. is_deleted — a post-hoc status not knowable at the decision moment for a live page.

## 3. Verify it with queries (grain, counts, missing values, windows)

*Every claim above gets a query cell here. A contract claim without a query next to it is a guess.*

Query A — grain check:

In [57]:
grain_check = con.sql(f"""
    SELECT content_hash_id, report_date, COUNT(*) AS n
    FROM '{BASE}/fact_content_daily_performance/month=2026-03/data_0.parquet'
    GROUP BY content_hash_id, report_date
    HAVING COUNT(*) > 1
    LIMIT 5
""").df()
print("Rows where (content_hash_id, report_date) is NOT unique:", len(grain_check))
grain_check

FloatProgress(value=0.0, layout=Layout(width='auto'), style=ProgressStyle(bar_color='black'))

Rows where (content_hash_id, report_date) is NOT unique: 0


,content_hash_id,report_date,n


Query B — row count + date span:

In [58]:
span = con.sql(f"""
    SELECT COUNT(*) AS n_rows,
           COUNT(DISTINCT content_hash_id) AS n_pages,
           MIN(report_date) AS min_date,
           MAX(report_date) AS max_date
    FROM '{BASE}/fact_content_daily_performance/month=2026-03/data_0.parquet'
""").df()
span

,n_rows,n_pages,min_date,max_date
0,9841378,331437,2026-03-01,2026-03-31


Query C — availability, filtered with IS TRUE:

In [59]:
avail = con.sql(f"""
    SELECT COUNT(*) AS total_rows,
           COUNT(*) FILTER (WHERE gsc_data_available IS TRUE) AS gsc_available_rows
    FROM '{BASE}/fact_content_daily_performance/month=2026-03/data_0.parquet'
""").df()
avail

FloatProgress(value=0.0, layout=Layout(width='auto'), style=ProgressStyle(bar_color='black'))

,total_rows,gsc_available_rows
0,9841378,3611061


February + March data aggregate

In [60]:
feb = con.sql(f"""
    SELECT content_hash_id, SUM(gsc_impressions) AS impressions_feb
    FROM '{BASE}/fact_content_daily_performance/month=2026-02/data_0.parquet'
    GROUP BY content_hash_id
""").df()

mar = con.sql(f"""
    SELECT content_hash_id, SUM(gsc_impressions) AS impressions_mar
    FROM '{BASE}/fact_content_daily_performance/month=2026-03/data_0.parquet'
    GROUP BY content_hash_id
""").df()

content_attrs = con.sql(f"""
    SELECT content_hash_id, word_count, content_type, search_volume, competition
    FROM '{BASE}/dim_content.parquet'
""").df()

print("Feb pages:", len(feb), "| March pages:", len(mar), "| Content attrs:", len(content_attrs))

FloatProgress(value=0.0, layout=Layout(width='auto'), style=ProgressStyle(bar_color='black'))

FloatProgress(value=0.0, layout=Layout(width='auto'), style=ProgressStyle(bar_color='black'))

FloatProgress(value=0.0, layout=Layout(width='auto'), style=ProgressStyle(bar_color='black'))

Feb pages: 321546 | March pages: 331437 | Content attrs: 519606


In [61]:
# Join Feb (features) + March (label source) + content attributes
lane = feb.merge(mar, on="content_hash_id", how="inner") \
          .merge(content_attrs, on="content_hash_id", how="inner")

# Label/proxy: declining if March impressions dropped vs February
lane["declining"] = (lane["impressions_mar"] < lane["impressions_feb"]).astype(int)

print("Lane rows after join:", len(lane))
print("Declining rate:", round(lane["declining"].mean(), 3))
lane[["content_hash_id", "impressions_feb", "impressions_mar", "word_count",
      "content_type", "search_volume", "competition", "declining"]].head()

Lane rows after join: 303572
Declining rate: 0.166


,content_hash_id,impressions_feb,impressions_mar,word_count,content_type,search_volume,competition,declining
0,content_1eea820697c3b95a,299.0,315.0,2613,keyword article,140,0.07,0
1,content_9abd8b303f805847,733.0,14536.0,2992,keyword article,90,0.00,0
2,content_5f58c55cbfee172a,514.0,387.0,2225,keyword article,720,0.00,1
3,content_6fe390ba3af1e456,2931.0,4697.0,2797,keyword article,110,0.00,0
4,content_3ad5d2160242b9ca,970.0,1004.0,2396,keyword article,70,0.00,0


In [62]:
from sklearn.tree import DecisionTreeClassifier
from sklearn.metrics import roc_auc_score

# Honest features — all knowable BEFORE March happens (from Feb data + static content attrs)
honest_features = ["impressions_feb", "word_count", "search_volume", "competition"]
X_honest = lane[honest_features].fillna(0)
y = lane["declining"]

tree_honest = DecisionTreeClassifier(max_depth=3, class_weight="balanced", random_state=42)
tree_honest.fit(X_honest, y)
score_honest = roc_auc_score(y, tree_honest.predict_proba(X_honest)[:, 1])
print(f"Honest baseline ROC AUC: {score_honest:.3f}")

Honest baseline ROC AUC: 0.849


In [63]:
# THE TRAP — add a feature computed FROM March data, which is exactly what we're predicting
lane["pct_change_leak"] = (lane["impressions_mar"] - lane["impressions_feb"]) / lane["impressions_feb"].replace(0, 1)

leaky_features = honest_features + ["pct_change_leak"]
X_leaky = lane[leaky_features].fillna(0)

tree_leaky = DecisionTreeClassifier(max_depth=3, class_weight="balanced", random_state=42)
tree_leaky.fit(X_leaky, y)
score_leaky = roc_auc_score(y, tree_leaky.predict_proba(X_leaky)[:, 1])
print(f"'Leaky' ROC AUC: {score_leaky:.3f}  <- looks amazing, but it's cheating")
print(f"\nWhy: pct_change_leak is calculated directly from impressions_mar — the same")
print(f"March data 'declining' is derived from. It's not a real feature, it's the answer in disguise.")

'Leaky' ROC AUC: 1.000  <- looks amazing, but it's cheating

Why: pct_change_leak is calculated directly from impressions_mar — the same
March data 'declining' is derived from. It's not a real feature, it's the answer in disguise.


In [64]:
# Remove the leak — this is the number that actually means something
lane = lane.drop(columns=["pct_change_leak"])
print(f"Kept: honest baseline ROC AUC = {score_honest:.3f} (not the fake {score_leaky:.3f})")

Kept: honest baseline ROC AUC = 0.849 (not the fake 1.000)


Five features (max), each knowable before the decision moment (before March 2026 starts):

1. impressions_feb — knowable at decision moment because it's fully observed by end of
   February, before we decide which pages to prioritize for March.
2. word_count — knowable because it's a static content attribute set when the page was
   published, unrelated to any future performance.
3. search_volume — knowable because it's a keyword-level metric independent of this
   page's actual performance.
4. competition — same as search_volume: a keyword attribute, not derived from outcomes.
5. content_type — knowable because it's fixed at publish time (e.g. "keyword article"),
   not something that changes based on performance.

Honest baseline (these 4 numeric features + content_type, via a depth-3 tree): ROC AUC 0.849.

## 4. Data limits

*What can this data never tell you? Unbalanced history, GSC-only early rows, window overlaps.*

This data can never tell you: (1) WHY a page declined — only that GSC impressions dropped,
not the cause (algorithm change, seasonality, a competitor, a technical issue). (2) The full
picture for every page — only 36.7% of March rows have gsc_data_available IS TRUE; the other
63% are missing GSC data for that day, which could bias any feature built from daily granularity
rather than the monthly sum used here. (3) Client history evenly — dim_clients shows an
unbalanced panel (gsc_data_start/ga4_data_start differ per client), so newer clients have
less history than older ones, and any model trained on this slice will underrepresent
newer clients. (4) Causation — a declining label is an observed correlation with time,
not proof that anything about the page itself caused the drop.

In [65]:
# Backing the "data limits" claims with numbers
print(f"GSC availability in March: {3611061/9841378:.1%} of rows")

unbalanced = con.sql(f"SELECT gsc_data_start, COUNT(*) as n_clients FROM '{BASE}/dim_clients.parquet' GROUP BY gsc_data_start ORDER BY n_clients DESC LIMIT 5").df()
print("\nClient history start dates vary (unbalanced panel):")
unbalanced


GSC availability in March: 36.7% of rows

Client history start dates vary (unbalanced panel):


,gsc_data_start,n_clients
0,NaT,37
1,2026-02-19,9
2,2025-09-24,7
3,2025-11-05,6
4,2025-10-20,2


## Self-check

Before you submit, confirm each line honestly:

- [x] Every section above is filled...
- [x] The notebook runs top to bottom...
- [x] No client names, URLs, or private queries anywhere
- [x] My claims use careful words: observed, measured, directional, decision-support
- [x] Committed to my repo under `work/notebooks/` — then submit your repo URL on the card. Done.